### DISCUSSION TITLE
Data Relations/Query Tracking

### DISCUSSION THREAD

**Discussion by Odigene12** [Ascend-Apollo-Apprentice-2024/learning-platform-api#25] [General]
Whenever we get all the students for a cohort from the following endpoint: `students?cohort={cohortId}`, the following JSON response comes back:

This is an example for cohortId = 15

```
{
    "count": 27,
    "next": null,
    "previous": null,
    "results": [
        {
            "id": 291,
            "name": "Gust Marks",
            "score": 0,
            "tags": [],
            "proposals": [],
            "book": {
                "id": 3,
                "name": "Book 2",
                "project": "Functionland"
            },
            "assessment_status": 0,
            "github": "jwwinburn",
            "cohorts": [
                {
                    "id": 15,
                    "name": "Day Cohort 63",
                    "start_date": "2023-02-13",
                    "end_date": "2023-08-04"
                }
            ],
            "archetype": ""
        }
    ]
}
```

**Comment by Odigene12** [comment]
The following are the datapoints that are used when creating the JSON.

student data:

`id` - `Ref: "LearningAPI_nssuser"."id"`

`name` - `SingleStudent` serializer -> `name` -> `get_name(self, obj)` (the serializer uses this method to get the data for name

`score` - `SingleStudent` serializer -> `score` -> `get_score(self, obj)` -> `student_score(obj)`  uses `LearningAPI_learningRecord` & `LearningAPI_CoreSkillRecord`

`tags` - `Ref: "LearningAPI_tag"."id" < "LearningAPI_nssusertag"."tag_id"`

**Reply by Odigene12** [reply]
## student score

 - **Concept:** The student score represents their level of competency in many concepts related to their learning.
 
### Tables Associated With Score

- LearningWeight: represents the many different learning objectives and their level of importance (weight)
- LearningRecord: captures whether or not a student achieved understanding of a competency and when.

`SELECT 
lr.student_id, 
SUM(lw.weight) AS score, 
lr.achieved
FROM "LearningAPI_learningrecord" AS lr
JOIN "LearningAPI_learningweight" as lw
ON lr.weight_id = lw.id
WHERE lr.student_id = 268
AND lr.achieved = TRUE
GROUP BY 
lr.student_id,
lr.achieved`

**Reply by Odigene12** [reply]
## name

- **Concept**: This represents the first and last name of a student.

### Tables Associated With Name

- NssUser: represents a one-to-one relationship with the auth_user table that gets generated by Django that we cannot touch.
- AuthUser: represents the Django generated table that holds a user's information such as `first_name`, `last_name`, etc.

**Reply by Odigene12** [reply]
## id

- **Concept**: The unique id of a a student in the NssUser table.

### Tables Associated with id

- NssUser: The id property comes from this table as its unique integer primary key.

**Reply by Odigene12** [reply]
## tags

- **Concept**: This represents a unique identifier name for students that is searchable in the UI that Instructors create.

### Tables Associated With Tags
- Tag: This table houses all the names of the tags that were created by the instructors.
- StudentTag: This is the join table that connects the tag and nss_user tables to show what students are associated with what tags.

**Reply by Odigene12** [reply]
## proposals

- **Concept:** This represents a collection of capstone proposals for each student that captures the capstone, capstone timeline, and proposal status.

### Tables Associated With Proposals
- Capstone: This table contains a URL path to a student's capstone proposal, a description of the capstone proposal, and the student and course associated with the capstone.
- Course: This is a table that captures the overarching courses in NSS. This has a foreign key to the capstone table for each course.
- NssUser: This has a foreign key to the capstone table for each student

**Reply by Odigene12** [reply]
## book

- **Concept:** This represents a section of a curriculum that students complete as part of their learning.

### Tables Associated With Book
- Book: This table contains the names and descriptions of tiles of parts of a curriculum that students take in a cohort.
- StudentProject: This table has a foreign key relationship with the NssUser and Project tables and tracks a student's progress.
- Project: This is a property on the StudentProject table and model that has a foreign key relationship with the Book table. This table contains the names of the projects that students are progressing through as part of a particular book.

**Comment by willkotheimer** [comment]
## Queries for Learning weights, Tags, and Name

```
SELECT  lr.student_id,  SUM(lw.weight) AS score
FROM "LearningAPI_learningrecord" AS lr 
JOIN "LearningAPI_learningweight" as lw ON lr.weight_id = lw.id 
WHERE lr.achieved = TRUE
GROUP BY  lr.student_id;

SELECT  lst.student_id, lst.tag_id, lt.name
from "LearningAPI_studenttag" AS lst
JOIN "LearningAPI_tag" as lt ON lt.id = lst.tag_id
WHERE  lst.tag_id IS NOT NULL;

SELECT nsu.user_id, CONCAT(au.first_name, ' ', au.last_name) as name
FROM "LearningAPI_nssuser" as nsu
JOIN "auth_user" as au ON au.id = nsu.user_id
```

### New concept: How serializers join:

Select queries on the model are "selected" by using the fields array, and models for the foreign keys are "included" or joined via the "depth" variable. Example: here the fields id and tag are used. Tag is a foreign key and depth = 1 causes the ORM to serialize the adjoining object in the response:

```class Meta:
        model = StudentTag
        fields = ('id', 'tag',)
        depth = 1
```

**Comment by jonathanvanduyne** [comment]
CTE for student

WITH student_score as (
    SELECT  lr.student_id,  SUM(lw.weight) AS score
    FROM "LearningAPI_learningrecord" AS lr 
    JOIN "LearningAPI_learningweight" as lw ON lr.weight_id = lw.id 
    WHERE lr.achieved = TRUE
    GROUP BY  lr.student_id 
),

 student_name as (
    SELECT nsu.user_id, CONCAT(au.first_name, ' ', au.last_name) as name
    FROM "LearningAPI_nssuser" as nsu
    JOIN "auth_user" as au ON au.id = nsu.user_id
),

 student_tag as (
    SELECT  lst.student_id, lst.tag_id, lt.name
    from "LearningAPI_studenttag" AS lst
    JOIN "LearningAPI_tag" as lt ON lt.id = lst.tag_id
    WHERE  lst.tag_id IS NOT NULL
)
SELECT ss.student_id, sn.name, ss.score, st.tag_id, st.name AS tag_name
FROM student_score AS ss
JOIN student_name AS sn ON ss.student_id = sn.user_id
LEFT JOIN student_tag AS st ON ss.student_id = st.student_id;

**Comment by Odigene12** [comment]
### First part of `get_proposals(self,obj)`

`proposals = Capstone.objects.filter(student=obj).annotate(
            status_count=Count("statuses"),
            approved=Count(
                'statuses',
                filter=Q(statuses__status__status="Approved")
            ),
            mvp=Count(
                'statuses',
                filter=Q(statuses__status__status="MVP")
            )
        ).order_by("pk")`

- `status_count=Count("statuses"):` Adds a field named `status_count` to each `Capstone` object, which contains the count of related `statuses `(assuming `Capstone `has a related field named `statuses`).
- `approved=Count('statuses', filter=Q(statuses__status__status="Approved")):` Adds a field named approved to each `Capstone` object, which contains the count of related `statuses `where the `status `field's `status `attribute is "Approved".
- `mvp=Count('statuses', filter=Q(statuses__status__status="MVP")):` Adds a field named `mvp `to each `Capstone `object, which contains the count of related `statuses `where the `status `field's `status `attribute is "MVP".
- The `.order_by("pk")` orders the annotated Capstone objects by their primary key (`pk`).

**Comment by jonathanvanduyne** [comment]
7/10/24 - Query to get students per cohort, adding in capstone data to continue tomorrow

SELECT u.id, cap.id as capstone_id
FROM "LearningAPI_nssuser" as u
JOIN "LearningAPI_nssusercohort" as uc ON u.id = uc.nss_user_id
LEFT JOIN "LearningAPI_capstone" as cap ON cap.student_id = u.id
WHERE uc.cohort_id = 15

**Comment by jonathanvanduyne** [comment]
7/11/22 - To Do next week: verify that new proposal entries return 'submitted'. Need to manipulate the data  as no students most likely are at that stage
 
SELECT 
    u.id, 
    cap.id as proposal_id, 
    co.id as course_id, 
    CASE
        WHEN ps.status = 'MVP' THEN 'mvp'
        WHEN ps.status = 'Approved' THEN 'approved'
        WHEN ps.status = 'In Review' THEN 'reviewed'
        WHEN ps.status IS NULL THEN 'submitted'
    END AS proposal_status

FROM "LearningAPI_nssuser" as u
JOIN "LearningAPI_nssusercohort" as uc ON u.id = uc.nss_user_id
LEFT JOIN "LearningAPI_capstone" as cap ON cap.student_id = u.id
LEFT JOIN "LearningAPI_capstonetimeline" as tl 
    on cap.id = tl.capstone_id
    and tl.id = (
        SELECT id
        FROM "LearningAPI_capstonetimeline"
        WHERE cap.id = capstone_id
        ORDER BY "date" DESC
        LIMIT 1
    )
LEFT JOIN "LearningAPI_proposalstatus" as ps on ps.id = tl.status_id
LEFT JOIN "LearningAPI_course" as co on co.id = cap.course_id 
WHERE uc.cohort_id = 9 
AND ps.status IS NOT NULL

**Comment by Odigene12** [comment]
Mayra is student 129 with 2 capstones

**Comment by willkotheimer** [comment]
We added a student capstone to the admin, and deleted all of her capstone_timelines, so that she would be only returning NULL on status. We still weren't seeing her though. This is because of the last line, `ps.status IS NOT NULL`. Since we wanted to show the submitted ones that were NULL, we removed that, and added an order by user id:

```
SELECT
tl."date",
u.id,
cap.id as proposal_id,
co.id as course_id,
CASE
WHEN ps.status = 'MVP' THEN 'mvp'
WHEN ps.status = 'Approved' THEN 'approved'
WHEN ps.status = 'In Review' THEN 'reviewed'
WHEN ps.status IS NULL THEN 'submitted'
END AS proposal_status

FROM "LearningAPI_nssuser" as u
JOIN "LearningAPI_nssusercohort" as uc ON u.id = uc.nss_user_id
LEFT JOIN "LearningAPI_capstone" as cap ON cap.student_id = u.id
LEFT JOIN "LearningAPI_capstonetimeline" as tl
on cap.id = tl.capstone_id
and tl.id = (
SELECT id
FROM "LearningAPI_capstonetimeline"
WHERE cap.id = capstone_id
ORDER BY "date" DESC
LIMIT 1
)
LEFT JOIN "LearningAPI_proposalstatus" as ps on ps.id = tl.status_id
LEFT JOIN "LearningAPI_course" as co on co.id = cap.course_id
WHERE uc.cohort_id = 9
Order by u.id
```

Here's the result set:

![image](https://github.com/user-attachments/assets/c3cec7f0-c649-4f64-8206-375fc16e906d)
